File: app/services/notification_service.py
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func, desc
from typing import Optional, List, Dict, Any
from datetime import datetime
import uuid

from app.models.notification import (
    Notification, NotificationTemplate, NotificationPreference,
    BroadcastMessage, NotificationType, NotificationStatus,
    NotificationCategory
)
from app.models.user import User
from app.schemas.notification import (
    NotificationCreate, NotificationUpdate,
    NotificationTemplateCreate, NotificationTemplateUpdate,
    SendNotificationRequest, NotificationDeliveryStatus,
    BroadcastMessageCreate
)
from fastapi import HTTPException, status

class NotificationChannel:
    def send_sms(self, phone: str, message: str) -> bool:
        return True

    def send_email(self, email: str, subject: str, message: str) -> bool:
        return True

    def send_push(self, user_id: int, title: str, message: str) -> bool:
        return True

class NotificationService:
    def __init__(self, db: Session):
        self.db = db
        self.channel = NotificationChannel()

    # Templates
    def create_template(self, data: NotificationTemplateCreate) -> NotificationTemplate:
        template = NotificationTemplate(
            name=data.name,
            title_template=data.title_template,
            message_template=data.message_template,
            notification_type=data.notification_type,
            category=data.category,
            is_active=True,
            variables=data.variables
        )
        self.db.add(template)
        self.db.commit()
        self.db.refresh(template)
        return template

    def update_template(self, template_id: int, data: NotificationTemplateUpdate) -> NotificationTemplate:
        template = self.db.query(NotificationTemplate).filter(NotificationTemplate.template_id == template_id).first()
        if not template:
            raise HTTPException(status_code=404, detail="Template not found")
        for k, v in data.dict(exclude_unset=True).items():
            setattr(template, k, v)
        self.db.commit()
        self.db.refresh(template)
        return template

    def render_template(self, name: str, variables: Dict[str, Any]) -> Dict[str, str]:
        template = self.db.query(NotificationTemplate).filter(
            and_(NotificationTemplate.name == name, NotificationTemplate.is_active == True)
        ).first()
        if not template:
            raise HTTPException(status_code=404, detail="Template not found")
        title = template.title_template.format(**variables)
        message = template.message_template.format(**variables)
        return {"title": title, "message": message}

    # Notifications
    def create_notification(self, data: NotificationCreate) -> Notification:
        user = self.db.query(User).filter(User.id == data.user_id).first()
        if not user:
            raise HTTPException(status_code=404, detail="User not found")
        notif = Notification(
            user_id=data.user_id,
            title=data.title,
            message=data.message,
            notification_type=data.notification_type,
            category=data.category,
            status=NotificationStatus.PENDING,
            priority=data.priority,
            scheduled_at=data.scheduled_at,
            metadata=data.metadata
        )
        self.db.add(notif)
        self.db.commit()
        self.db.refresh(notif)
        return notif

    def update_notification(self, notification_id: int, data: NotificationUpdate) -> Notification:
        notif = self.db.query(Notification).filter(Notification.notification_id == notification_id).first()
        if not notif:
            raise HTTPException(status_code=404, detail="Notification not found")
        for k, v in data.dict(exclude_unset=True).items():
            setattr(notif, k, v)
        self.db.commit()
        self.db.refresh(notif)
        return notif

    def send_notification_now(self, notification_id: int) -> NotificationDeliveryStatus:
        notif = self.db.query(Notification).filter(Notification.notification_id == notification_id).first()
        if not notif:
            raise HTTPException(status_code=404, detail="Notification not found")
        user = self.db.query(User).filter(User.id == notif.user_id).first()
        if not user:
            raise HTTPException(status_code=404, detail="User not found")
        success = False
        if notif.notification_type == NotificationType.SMS and user.phone:
            success = self.channel.send_sms(user.phone, notif.message)
        elif notif.notification_type == NotificationType.EMAIL and user.email:
            success = self.channel.send_email(user.email, notif.title, notif.message)
        elif notif.notification_type in [NotificationType.PUSH, NotificationType.IN_APP]:
            success = self.channel.send_push(user.id, notif.title, notif.message)
        notif.sent_at = datetime.utcnow()
        notif.status = NotificationStatus.SENT if success else NotificationStatus.FAILED
        self.db.commit()
        return NotificationDeliveryStatus(
            notification_id=notif.notification_id,
            status=notif.status,
            sent_at=notif.sent_at,
            delivered_at=None,
            read_at=None,
            error_message=None if success else "Delivery failed"
        )

    def send_bulk(self, req: SendNotificationRequest) -> Dict[str, Any]:
        created, sent, failed = 0, 0, 0
        errors: List[str] = []
        for user_id in req.user_ids:
            title = req.title or ""
            message = req.message or ""
            if req.template_name:
                try:
                    rendered = self.render_template(req.template_name, req.template_variables or {})
                    title = title or rendered["title"]
                    message = message or rendered["message"]
                except Exception as e:
                    errors.append(f"Template render failed for user {user_id}: {e}")
                    failed += 1
                    continue
            notif = self.create_notification(NotificationCreate(
                user_id=user_id,
                title=title,
                message=message,
                notification_type=req.notification_type,
                category=req.category,
                priority=req.priority,
                scheduled_at=req.scheduled_at,
                metadata=req.metadata
            ))
            created += 1
            # For demo, send immediately if not scheduled
            if not req.scheduled_at:
                status_resp = self.send_notification_now(notif.notification_id)
                sent += 1 if status_resp.status in [NotificationStatus.SENT, NotificationStatus.DELIVERED] else 0
        return {
            "success": failed == 0,
            "notifications_created": created,
            "notifications_sent": sent,
            "failed_count": failed,
            "errors": errors
        }

    # Broadcasts
    def create_broadcast(self, data: BroadcastMessageCreate, created_by: int) -> BroadcastMessage:
        b = BroadcastMessage(
            title=data.title,
            message=data.message,
            notification_type=data.notification_type,
            target_audience=data.target_audience,
            scheduled_at=data.scheduled_at,
            created_by=created_by
        )
        self.db.add(b)
        self.db.commit()
        self.db.refresh(b)
        return b

    def execute_broadcast(self, broadcast_id: int) -> Dict[str, Any]:
        b = self.db.query(BroadcastMessage).filter(BroadcastMessage.broadcast_id == broadcast_id).first()
        if not b:
            raise HTTPException(status_code=404, detail="Broadcast not found")
        # Determine audience
        query = self.db.query(User)
        if b.target_audience == "drivers":
            query = query.filter(User.role == "driver")
        elif b.target_audience == "customers":
            query = query.filter(User.role == "customer")
        elif b.target_audience == "admins":
            query = query.filter(User.role == "admin")
        users = query.all()
        b.total_recipients = len(users)
        sent = 0
        failed = 0
        for user in users:
            notif = self.create_notification(NotificationCreate(
                user_id=user.id,
                title=b.title,
                message=b.message,
                notification_type=b.notification_type,
                category=NotificationCategory.BROADCAST,
                priority="high"
            ))
            status_resp = self.send_notification_now(notif.notification_id)
            if status_resp.status in [NotificationStatus.SENT, NotificationStatus.DELIVERED]:
                sent += 1
            else:
                failed += 1
        b.sent_count = sent
        b.failed_count = failed
        b.sent_at = datetime.utcnow()
        self.db.commit()
        return {"broadcast_id": b.broadcast_id, "sent": sent, "failed": failed}

    # Stats
    def get_statistics(self) -> Dict[str, Any]:
        total = self.db.query(Notification).count()
        sent = self.db.query(Notification).filter(Notification.status == NotificationStatus.SENT).count()
        delivered = self.db.query(Notification).filter(Notification.status == NotificationStatus.DELIVERED).count()
        failed = self.db.query(Notification).filter(Notification.status == NotificationStatus.FAILED).count()
        read = self.db.query(Notification).filter(Notification.status == NotificationStatus.READ).count()
        by_type = {
            t.value: self.db.query(Notification).filter(Notification.notification_type == t).count()
            for t in NotificationType
        }
        by_category = {
            c.value: self.db.query(Notification).filter(Notification.category == c).count()
            for c in NotificationCategory
        }
        delivery_rate = (delivered / total * 100) if total else 0.0
        read_rate = (read / total * 100) if total else 0.0
        return {
            "total_notifications": total,
            "sent_notifications": sent,
            "delivered_notifications": delivered,
            "failed_notifications": failed,
            "read_notifications": read,
            "delivery_rate": delivery_rate,
            "read_rate": read_rate,
            "notifications_by_type": by_type,
            "notifications_by_category": by_category
        }

File: src/services/notificationService.ts
// Notification API Service
// This file contains all API calls for notification management functionality

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

// Define interfaces for notifications
interface Notification {
  notification_id?: number;
  title: string;
  message: string;
  notification_type: 'sms' | 'email' | 'push';
  category?: string;
  priority: 'low' | 'normal' | 'high';
  user_id?: number;
  status: 'pending' | 'sent' | 'delivered' | 'failed' | 'cancelled';
  scheduled_at?: string;
  sent_at?: string;
  delivered_at?: string;
  read_at?: string;
  metadata?: Record<string, unknown>;
  retry_count?: number;
  error_message?: string;
  created_at?: string;
  updated_at?: string;
}

// Define interface for notification templates
interface NotificationTemplate {
  template_id?: number;
  name: string;
  title_template: string;
  message_template: string;
  notification_type: 'sms' | 'email' | 'push';
  category: string;
  is_active: boolean;
  variables?: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
}

// Define interface for notification list response
interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  page: number;
  size: number;
  unread_count: number;
}

// Define interface for notification stats response
interface NotificationStatsResponse {
  total_notifications: number;
  sent_notifications: number;
  delivered_notifications: number;
  failed_notifications: number;
  read_notifications: number;
  delivery_rate: number;
  read_rate: number;
  notifications_by_type: Record<string, number>;
  notifications_by_category: Record<string, number>;
}

// Define interfaces for broadcast notifications
interface BroadcastNotification {
  broadcast_id?: number;
  title: string;
  message: string;
  notification_type: 'sms' | 'email' | 'push';
  target_audience: string;
  scheduled_at?: string;
  sent_at?: string;
  total_recipients?: number;
  sent_count?: number;
  failed_count?: number;
  created_by: number;
  created_at?: string;
}

// Define interfaces for template-based notifications
interface TemplateNotificationRequest {
  user_ids: number[];
  template_name: string;
  title?: string;
  message?: string;
  notification_type: 'sms' | 'email' | 'push';
  category?: string;
  priority?: 'low' | 'normal' | 'high';
  scheduled_at?: string;
  template_variables?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

interface TemplateNotificationResponse {
  success: boolean;
  notifications_created: number;
  notifications_sent: number;
  failed_count: number;
  errors: string[];
}

interface SendNotificationResponse {
  notification_id: number;
  status: string;
  sent_at: string;
  delivered_at: string | null;
  read_at: string | null;
  error_message: string | null;
}

interface ExecuteBroadcastResponse {
  broadcast_id: number;
  sent: number;
  failed: number;
}

interface PaginatedResponse<T> {
  data: T[];
  total?: number;
  page?: number;
  limit?: number;
  total_pages?: number;
}

interface FilterOptions {
  category?: string;
  notification_type?: string;
  status?: string;
  priority?: string;
  user_id?: number;
  search?: string;
  page?: number;
  limit?: number;
}

class NotificationService {
  private getAuthHeaders(): HeadersInit {
    const token = localStorage.getItem('access_token');
    return {
      'Content-Type': 'application/json',
      'Authorization': token ? `Bearer ${token}` : '',
    };
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    if (!response.ok) {
      if (response.status === 401 || response.status === 403) {
        localStorage.removeItem('access_token');
        localStorage.removeItem('current_user');
        window.location.href = '/auth/signin';
        throw new Error('Authentication failed. Please log in again.');
      }
      
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
    }
    return response.json();
  }

  // Get all notifications with pagination and filters
  async getNotifications(filters: FilterOptions = {}): Promise<PaginatedResponse<Notification>> {
    const params = new URLSearchParams();
    
    // Add filters
    if (filters.category) params.append('category', filters.category);
    if (filters.notification_type) params.append('notification_type', filters.notification_type);
    if (filters.status) params.append('status', filters.status);
    if (filters.priority) params.append('priority', filters.priority);
    if (filters.user_id) params.append('user_id', filters.user_id.toString());
    if (filters.search) params.append('search', filters.search);
    
    // Add pagination parameters
    if (filters.page && filters.page > 0) {
      const skip = (filters.page - 1) * (filters.limit || 10);
      params.append('skip', skip.toString());
    }
    
    // Add limit with validation
    if (filters.limit) {
      const limit = Math.min(Math.max(filters.limit, 1), 100);
      params.append('limit', limit.toString());
    } else {
      params.append('limit', '10'); // Default limit
    }

    const response = await fetch(`${API_BASE_URL}/notifications/?${params.toString()}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });
    
    const notifications = await this.handleResponse<Notification[]>(response);
    return { data: notifications };
  }

  // Get notification by ID
  async getNotificationById(notificationId: number): Promise<Notification> {
    const response = await fetch(`${API_BASE_URL}/notifications/${notificationId}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<Notification>(response);
  }

  // Create new notification
  async createNotification(notificationData: Omit<Notification, 'notification_id' | 'created_at' | 'updated_at' | 'status' | 'sent_at' | 'delivered_at' | 'read_at' | 'retry_count' | 'error_message'>): Promise<Notification> {
    const response = await fetch(`${API_BASE_URL}/notifications`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(notificationData),
    });

    return this.handleResponse<Notification>(response);
  }

  // Update notification
  async updateNotification(notificationId: number, notificationData: Partial<Notification>): Promise<Notification> {
    const response = await fetch(`${API_BASE_URL}/notifications/${notificationId}`, {
      method: 'PUT',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(notificationData),
    });

    return this.handleResponse<Notification>(response);
  }

  // Delete notification
  async deleteNotification(notificationId: number): Promise<{ message: string }> {
    const response = await fetch(`${API_BASE_URL}/notifications/${notificationId}`, {
      method: 'DELETE',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
    }

    return response.json();
  }

  // Send notification immediately
  async sendNotification(notificationId: number): Promise<SendNotificationResponse> {
    const response = await fetch(`${API_BASE_URL}/notifications/${notificationId}/send`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<SendNotificationResponse>(response);
  }

  // Cancel scheduled notification
  async cancelNotification(notificationId: number): Promise<Notification> {
    const response = await fetch(`${API_BASE_URL}/notifications/${notificationId}/cancel`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<Notification>(response);
  }

  // Send template-based notifications to multiple users
  async sendTemplateNotifications(requestData: TemplateNotificationRequest): Promise<TemplateNotificationResponse> {
    const response = await fetch(`${API_BASE_URL}/notifications/send`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(requestData),
    });

    return this.handleResponse<TemplateNotificationResponse>(response);
  }

  // Get all broadcast notifications
  async getBroadcastNotifications(): Promise<BroadcastNotification[]> {
    const response = await fetch(`${API_BASE_URL}/notifications/broadcasts`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<BroadcastNotification[]>(response);
  }

  // Create new broadcast notification
  async createBroadcastNotification(broadcastData: Omit<BroadcastNotification, 'broadcast_id' | 'created_at' | 'sent_at' | 'total_recipients' | 'sent_count' | 'failed_count'>, created_by: number): Promise<BroadcastNotification> {
    const params = new URLSearchParams();
    params.append('created_by', created_by.toString());
    
    const response = await fetch(`${API_BASE_URL}/notifications/broadcasts?${params.toString()}`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(broadcastData),
    });

    return this.handleResponse<BroadcastNotification>(response);
  }

  // Get broadcast notification by ID
  async getBroadcastNotificationById(broadcastId: number): Promise<BroadcastNotification> {
    const response = await fetch(`${API_BASE_URL}/notifications/broadcasts/${broadcastId}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<BroadcastNotification>(response);
  }

  // Update broadcast notification
  async updateBroadcastNotification(broadcastId: number, broadcastData: Partial<BroadcastNotification>): Promise<BroadcastNotification> {
    const response = await fetch(`${API_BASE_URL}/notifications/broadcasts/${broadcastId}`, {
      method: 'PUT',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(broadcastData),
    });

    return this.handleResponse<BroadcastNotification>(response);
  }

  // Delete broadcast notification
  async deleteBroadcastNotification(broadcastId: number): Promise<{ message: string }> {
    const response = await fetch(`${API_BASE_URL}/notifications/broadcasts/${broadcastId}`, {
      method: 'DELETE',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
    }

    return response.json();
  }

  // Execute broadcast notification (send to all recipients)
  async executeBroadcast(broadcastId: number): Promise<ExecuteBroadcastResponse> {
    const response = await fetch(`${API_BASE_URL}/notifications/broadcasts/${broadcastId}/execute`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<ExecuteBroadcastResponse>(response);
  }

  // Get notifications list with pagination (new endpoint)
  async getNotificationsList(skip: number = 0, limit: number = 50): Promise<NotificationListResponse> {
    const params = new URLSearchParams();
    params.append('skip', skip.toString());
    params.append('limit', limit.toString());
    
    const response = await fetch(`${API_BASE_URL}/notifications/list?${params.toString()}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });
    
    return this.handleResponse<NotificationListResponse>(response);
  }

  // Get notification stats (new endpoint)
  async getNotificationStats(): Promise<NotificationStatsResponse> {
    const response = await fetch(`${API_BASE_URL}/notifications/stats`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });
    
    return this.handleResponse<NotificationStatsResponse>(response);
  }

  // Template methods
  async getTemplates(filters: FilterOptions = {}): Promise<PaginatedResponse<NotificationTemplate>> {
    const params = new URLSearchParams();
    
    // Add filters
    if (filters.category) params.append('category', filters.category);
    if (filters.notification_type) params.append('notification_type', filters.notification_type);
    if (filters.search) params.append('search', filters.search);
    
    // Add pagination parameters
    if (filters.page && filters.page > 0) {
      const skip = (filters.page - 1) * (filters.limit || 10);
      params.append('skip', skip.toString());
    }
    
    // Add limit with validation
    if (filters.limit) {
      const limit = Math.min(Math.max(filters.limit, 1), 100);
      params.append('limit', limit.toString());
    } else {
      params.append('limit', '10'); // Default limit
    }

    const response = await fetch(`${API_BASE_URL}/notifications/templates?${params.toString()}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });
    
    const templates = await this.handleResponse<NotificationTemplate[]>(response);
    return { data: templates };
  }

  async getTemplateById(templateId: number): Promise<NotificationTemplate> {
    const response = await fetch(`${API_BASE_URL}/notifications/templates/${templateId}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    return this.handleResponse<NotificationTemplate>(response);
  }

  async createTemplate(templateData: Omit<NotificationTemplate, 'template_id' | 'created_at' | 'updated_at'>): Promise<NotificationTemplate> {
    const response = await fetch(`${API_BASE_URL}/notifications/templates`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(templateData),
    });

    return this.handleResponse<NotificationTemplate>(response);
  }

  async updateTemplate(templateId: number, templateData: Partial<NotificationTemplate>): Promise<NotificationTemplate> {
    const response = await fetch(`${API_BASE_URL}/notifications/templates/${templateId}`, {
      method: 'PUT',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(templateData),
    });

    return this.handleResponse<NotificationTemplate>(response);
  }

  async deleteTemplate(templateId: number): Promise<{ message: string }> {
    const response = await fetch(`${API_BASE_URL}/notifications/templates/${templateId}`, {
      method: 'DELETE',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
    }

    return response.json();
  }
}

// Export singleton instance
export const notificationService = new NotificationService();
export type { Notification, BroadcastNotification, NotificationTemplate, TemplateNotificationRequest, TemplateNotificationResponse, SendNotificationResponse, ExecuteBroadcastResponse, PaginatedResponse, FilterOptions, NotificationListResponse, NotificationStatsResponse };